### DISCUSSION TITLE
What's wrong with this snippet?

### DISCUSSION THREAD

**Discussion by danielo515** [L3MON4D3/LuaSnip#946] [Q&A] [answered]
I swear I tried every possible fix, but I can't figure out what is wrong with this:
```lua
  s(
    { trig = "enum", dscr = "complete enum" },
    fmta([[
    #[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord, Display)]
    pub enum <> {
        <>
    }
    ]], i(1), i(0))
  ),
```
It keeps complaining about
```
Missing key `2` in format arguments:
```
What key 2? I only have two format arguments, what is this thing missing?

**Comment by danielo515** [comment, marked as answer]
I can't express how angry I am right now... turns out the problem whas that arguments must be supplied in a table.
I copy pasted previous snippets that only had one argument, and, because only Lua gods know why, it was working, but as soon as it requires more than one argument, it fails miserably. Sometimes I wish lua was not so weak in terms of function arguments. An error saying "You must provide a table as second argument" would have been soooo much more useful...

**Reply by L3MON4D3** [reply]
Man, sorry about that :/
This "issue" is with us though, if there is only a single node passed, we wrap it in a table.
In hindsight, maybe not the best idea, since it causes issues like yours

**Reply by danielo515** [reply]
Don't worry, you wrote an awesome snippet engine, whose only flaw is be written in Lua 😁
I was trying to avoid writing Haxe externs for Luasnip, but maybe it will be worth it.
Thanks mate

**Reply by L3MON4D3** [reply]
What are haxe externs? I have not heard about that yet (feel free to reply with some link :D)

**Reply by danielo515** [reply]
You asked me THE question 😄 
[Haxe](https://haxe.org/) is a super convenient programming language, mostly used to create videogames. It's main strengths are (in my oppinion) two:

- it's great type system. If you want to compare it with something, is as good as rust one, but a lot more flexible. Both languages took inspiration from Ocaml, so they share a lot of similarities. For example, both have the ability too attach new functions to any existing type (traits in rust, static extensions in Haxe).
- it was designed as a compile to other languages from the ground up. This made very easy to compile Haxe to a bunch of other languages. The list of targets is impressinve, and keeps growing, to name a few: C++, JavaScript, Lua, C#, Java...

After wasting hours trying to debug stupid mistakes I was making in Lua I decided to find a typed language that I can use to write neovim plugins, and Haxe was the perfect fit.
Obviously, the majority of people use it for creating videogames, but the community is super welcoming and helped me a lot creating a toolchain to write Neovim Plugins using Lua. Here is the main repo of the project:
https://github.com/danielo515/haxe-nvim
And here is one example use case ( my personal neovim configuration):
https://github.com/danielo515/kickstart.hx

**Reply by L3MON4D3** [reply]
Oh that's cool, composability > inheritance any day :D
Compiling to other languages also seems interesting, it's primarily done to easily integrate into existing codebases, yes?
I do have to say, Haxe has a bit more structure than lua, which I feel like might get in the way when configuring neovim (just because there is more to type), but when writing plugins it's probably a boon, at least once they get as complex as luasnip (which is positively sprawling, and not as cleanly structured as I'd like :sweat_smile:)

Looks like a great effort on your part, hope it becomes more popular :crossed_fingers: :D

**Reply by danielo515** [reply]
> composability > inheritance any day :D

Yeah, indeed. Why you mention it? Because Haxe? Don't get fooled (like I was) by the first impression the language gives. It's main aim is video game programming, and as much as I love functional programming, for video games OOP makes sense (probably the only case where it does, LOL) hence they showcase a lot of OOP features. But they are more functional capable than many other languages (more than Java, C, or C#). In fact, I only write static functions and only use clases as namespace, very much like in Lua. And their macro system is pure love.
I'm not trying to convince you to use it 😅 (but if you want, be my guest, haha) I'm just being fair to the language.
I don't want to leave the opportunity to thank you again for Luasnip, the best snippet engine ever built.

**Reply by L3MON4D3** [reply]
Mentioning it because you saif that haxe externs are like rust traits, and those being more comfortable than regular inheritance is something I can relate to :D
Thanks for telling me a bit about Haxe, appreciate it! If I ever want to do more with it, I know where to get some help :D
Thanks for the thanks xD I'm very happy you like it, and hope you enjoy using it, beside minor snags (like above) :D
